chainstate/test-suite/src/tests/orders_tests.rs

```diff
@@ -0,0 +1,664 @@
+fn fill_order_check_storage(#[case] seed: Seed) {
+    utils::concurrency::model(move || {
+        let mut rng = make_seedable_rng(seed);
+        let mut tf = TestFramework::builder(&mut rng).build();
+
+        let (token_id, tokens_outpoint, coins_outpoint) = issue_and_mint_token(&mut rng, &mut tf);
+
+        let ask_amount = Amount::from_atoms(rng.gen_range(10u128..1000));
+        let give_amount = Amount::from_atoms(rng.gen_range(10u128..1000));
+        let order_data = OrderData::new(
+            Destination::AnyoneCanSpend,
+            OutputValue::Coin(ask_amount),
+            OutputValue::TokenV1(token_id, give_amount),
+        );
+
+        let order_id = make_order_id(&tokens_outpoint);
+        let tx = TransactionBuilder::new()
+            .add_input(tokens_outpoint.into(), InputWitness::NoSignature(None))
+            .add_output(TxOutput::AnyoneCanTake(order_data.clone()))
+            .build();
+        tf.make_block_builder().add_transaction(tx).build_and_process(&mut rng).unwrap();
+
+        // Fill the order partially
+        let fill_value = OutputValue::Coin(Amount::from_atoms(
+            rng.gen_range(1..ask_amount.into_atoms()),
+        ));
+        let filled_amount = {
+            let db_tx = tf.storage.transaction_ro().unwrap();
+            let orders_db = OrdersAccountingDB::new(&db_tx);
+            orders_accounting::calculate_fill_order(&orders_db, order_id, &fill_value).unwrap()
+        };
+        let left_to_fill = (ask_amount - fill_value.amount()).unwrap();
+
+        let tx = TransactionBuilder::new()
+            .add_input(coins_outpoint.into(), InputWitness::NoSignature(None))
+            .add_input(
+                TxInput::AccountCommand(
+                    AccountNonce::new(0),
+                    AccountCommand::FillOrder(order_id, fill_value, Destination::AnyoneCanSpend),
+                ),
+                InputWitness::NoSignature(None),
+            )
+            .add_output(TxOutput::Transfer(
+                OutputValue::TokenV1(token_id, filled_amount),
+                Destination::AnyoneCanSpend,
+            ))
+            .add_output(TxOutput::Transfer(
+                OutputValue::Coin(left_to_fill),
+                Destination::AnyoneCanSpend,
+            ))
+            .build();
+        let partial_fill_tx_id = tx.transaction().get_id();
+        tf.make_block_builder().add_transaction(tx).build_and_process(&mut rng).unwrap();
+
+        assert_eq!(
+            Some(order_data),
+            tf.chainstate.get_order_data(&order_id).unwrap()
+        );
+        assert_eq!(
+            Some(left_to_fill),
+            tf.chainstate.get_order_ask_balance(&order_id).unwrap()
+        );
+        assert_eq!(
+            Some((give_amount - filled_amount).unwrap()),
+            tf.chainstate.get_order_give_balance(&order_id).unwrap()
+        );
+
+        // Fill the rest of the order
+        let fill_value = OutputValue::Coin(left_to_fill);
```

Comment:
There should be tests that attempt `fill_value + 1` and `fill_value - 1`

Reply:
these tests are in chainstate/constraints-value-accumulator/src/tests/orders_constraints.rs